pkg/lib/manifest2po

```diff
@@ -0,0 +1,107 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import argparse
+import collections
+import collections.abc
+import json
+import pathlib
+from typing import Any
+
+PO_HEADER = """msgid ""
+msgstr ""
+"Project-Id-Version: PACKAGE_VERSION\\n"
+"MIME-Version: 1.0\\n"
+"Content-Type: text/plain; charset=UTF-8\\n"
+"Content-Transfer-Encoding: 8bit\\n"
+"X-Generator: Cockpit manifest2po\\n"
+"""
+
+
+def get_docs_strings(docs: list[dict[str, str]]) -> collections.abc.Iterable[str]:
+    for doc in docs:
+        assert 'label' in doc, 'doc entry without label'
+        yield doc['label']
+
+
+def get_keyword_strings(keywords: list[dict[str, list[str]]]) -> collections.abc.Iterable[str]:
+    for keyword in keywords:
+        assert 'matches' in keyword, 'keywords entry without matches'
+        for match in keyword['matches']:
+            yield match
+
+
+def get_menu_strings(menu: dict[str, Any]) -> collections.abc.Iterable[str]:
+    for entry in menu.values():
+        if label := entry.get('label'):
+            yield label
+        if keywords := entry.get('keywords'):
+            yield from get_keyword_strings(keywords)
+        if docs := entry.get('docs'):
+            yield from get_docs_strings(docs)
+
+
+def get_bridges_strings(bridges: list[dict[str, str]]) -> collections.abc.Iterable[str]:
+    for bridge in bridges:
+        if label := bridge.get('label'):
+            yield label
+
+
+def get_manifest_strings(manifest: dict[str, Any]) -> collections.abc.Iterable[str]:
+    if menu := manifest.get('menu'):
+        yield from get_menu_strings(menu)
+    if tools := manifest.get('tools'):
+        yield from get_menu_strings(tools)
+    if bridges := manifest.get('bridges'):
+        yield from get_bridges_strings(bridges)
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(prog='manifest2po',
+                                     description='Extracts translatable strings from manifest.json files')
+    parser.add_argument('-d', '--directory', help='Base directory for input files')
+    parser.add_argument('-o', '--output', help='Output files', required=True)
+    parser.add_argument('files', nargs='+', help='One or more input files', type=pathlib.Path, metavar='FILE')
+
+    args = parser.parse_args()
+    strings = collections.defaultdict[str, set[str]](set)
+
+    file: pathlib.Path
+    for file in args.files:
+        if file.name != 'manifest.json':
+            continue
+
+        # Qualify the filename if necessary
+        full_path = args.directory / file if args.directory else file
```

Comment:
This strikes me as sort of odd.... are we just doing that because we want to get the correct relative paths in the comments in the output?

Reply:
I suppose? We do:

```
mkdir -p po/ && \
./pkg/lib/html2po.js -d . -o po/cockpit.html.pot \
        $(cd . && find pkg/ -name '*.html')
```